numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -3286,9 +3287,19 @@ PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
     if (common_dt == NULL) {
         goto fail;
     }
+    PyArray_Descr *x_dt, *y_dt;
+    int has_ref = PyDataType_REFCHK(common_dt);
+    if (has_ref) {
+        x_dt = PyArray_DESCR(op_in[2]);
+        y_dt = PyArray_DESCR(op_in[3]);
+    }
+    else {
+        x_dt = common_dt;
+        y_dt = common_dt;
+    }
```

Comment:
For `has_ref` we do an explicit cast and don't let the iterator handle the cast for us. Otherwise, the iterator handles the cast.

Maybe I should add the above text as a comment?

Reply:
A comment would make sense I think.  In principle, whenever we don't use the fast-path below, doing this probably makes sense (not casting in the iterator).

I am not sure that "has-ref" is in practice the right check, or whether it shouldn't be the "may create a new dtype on copy" (although in practice has-ref covers that).

Reply:
With a bit clearer head, maybe let me clarify:

My comment about checking for "may create a new dtype on copy" seems incorrect.  We assume that `has_ref` is sufficient to say that we are allowed to trivially copy the data.
I think it is safe to say that a dtype which does spawn new instance here must ensure that trivial copies are still allowed so that we are fine.
(For fresh clones of the dtype anyway, one could imagine mutable categoricals.)

I still think if it doesn't clutter things: it would be nice to fully align casting to `common_dt` to whether we can use the trivial-copy fast paths or not.
But as I said, I am happy with just a comment explaining it.

Reply:
Thanks for the update, about the strides, they really doen't matter, but it is a bit nicer to fix maybe.

I think your old logic of checking the `has_ref` on the common type makes a bit more sense actually?  since that is what we are arguing about being a "trivial copy"?

If anything, I was wondering about moving a:
```
int no_trivial_loop = has_ref || ((itemsize != 16) && (itemsize != 8) && (itemsize != 4) &&
                    (itemsize != 2) && (itemsize != 1))
```
to early on to reuse it.

In either case, this is good though!